core/node/api_server/src/utils.rs

```diff
@@ -261,6 +266,103 @@ async fn fill_receipts_with_unknown_nonce(
     Ok(())
 }
 
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum AccountType {
+    /// Externally owned account.
+    External(ExternalAccountType),
+    /// Non-AA contract.
+    Contract,
+}
+
+impl AccountType {
+    fn is_default(self) -> bool {
+        matches!(self, Self::External(ExternalAccountType::Default))
+    }
+
+    pub(crate) fn is_external(self) -> bool {
+        matches!(self, Self::External(_))
+    }
+}
+
+impl AccountType {
+    async fn with_full_nonce(
+        storage: &mut Connection<'_, Core>,
+        address: Address,
+        block_number: L2BlockNumber,
+    ) -> Result<(Self, U256), Web3Error> {
+        let code_key = get_code_key(&address).hashed_key();
+        let is_account_key = get_is_account_key(&address).hashed_key();
+        let nonce_key = get_nonce_key(&address).hashed_key();
+
+        let values = storage
+            .storage_logs_dal()
+            .get_storage_values(&[nonce_key, code_key, is_account_key], block_number)
+            .await
+            .map_err(DalError::generalize)?;
+        let full_nonce = values[&nonce_key].unwrap_or_default();
+        let code_hash = values[&code_key].unwrap_or_default();
+        let account_info = values[&is_account_key].unwrap_or_default();
+
+        let ty = if code_hash.is_zero() {
+            Self::External(ExternalAccountType::Default)
+        } else if account_info.is_zero() {
+            Self::Contract
+        } else {
+            Self::External(ExternalAccountType::Custom)
+        };
+        Ok((ty, h256_to_u256(full_nonce)))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct AccountTypesCache {
+    cache: LruCache<Address, AccountType>,
```

Comment:
Maybe, having a cache is overcomplicating things? I'm reasonably sure it works correctly, and I expect it to be useful. OTOH, if we can afford having 3 storage accesses (within a single Postgres query) on every `eth_getTransactionCount` call, the additional complexity can be avoided.

Reply:
I wouldn't add a cache but having is also ok with me